File: src/sync_planner.rb
# frozen_string_literal: true

require "set"
require_relative "label_identity"
require_relative "sync_plan"

module SyncLabels
  class SyncPlanner
    def initialize(config:)
      @config = config
    end

    def plan(existing)
      entries = []
      labels_by_name = existing.to_h { |label| [LabelIdentity.key(label["name"]), label] }
      desired_keys = @config.labels.map { |label| LabelIdentity.key(label["name"]) }.to_set

      @config.labels.each do |desired|
        desired_name = desired["name"]
        desired_key = LabelIdentity.key(desired_name)
        current = labels_by_name[desired_key]
        alias_matches = desired["aliases"].map { |name| labels_by_name[LabelIdentity.key(name)] }.compact
          .uniq { |label| LabelIdentity.key(label["name"]) }

        if current
          action = label_changed?(current, desired) ? :update : :unchanged
          entries << entry(action, current.fetch("name"), desired: action == :update ? desired : nil)
          labels_by_name.delete(LabelIdentity.key(current["name"]))
          labels_by_name[desired_key] = desired

          alias_matches.each do |legacy|
            entries << entry(:delete, legacy.fetch("name"), reason: :legacy_alias)
            labels_by_name.delete(LabelIdentity.key(legacy["name"]))
          end
          next
        end

        if alias_matches.length > 1
          names = alias_matches.map { |label| label["name"] }.join(", ")
          raise "多个旧标签同时映射到 #{desired_name}：#{names}"
        end

        if alias_matches.length == 1
          old = alias_matches.first
          entries << entry(:rename, old.fetch("name"), desired: desired)
          labels_by_name.delete(LabelIdentity.key(old["name"]))
          labels_by_name[desired_key] = desired
          next
        end

        entries << entry(:create, desired_name, desired: desired)
        labels_by_name[desired_key] = desired
      end

      remaining = labels_by_name.values.reject { |label| desired_keys.include?(LabelIdentity.key(label["name"])) }
      stale_managed, repository_specific = remaining.partition { |label| @config.managed?(label["name"]) }

      stale_managed.sort_by { |label| label["name"].downcase }.each do |label|
        entries << entry(:delete, label.fetch("name"), reason: :stale_managed)
      end
      repository_specific.sort_by { |label| label["name"].downcase }.each do |label|
        entries << entry(:preserve, label.fetch("name"))
      end

      SyncPlan.new(entries: entries)
    end

    private

    def label_changed?(current, desired)
      current["name"] != desired["name"] ||
        current["color"].to_s.upcase != desired["color"] ||
        current["description"].to_s != desired["description"]
    end

    def entry(action, name, desired: nil, reason: nil)
      PlanEntry.new(action: action, name: name, desired: desired, reason: reason)
    end
  end
end

File: src/sync-planner.ts
import { labelKey } from "./label-identity";
import type { ExistingLabel, LabelDefinition, PlanningConfig } from "./label-types";
import { SyncPlan, type PlanEntry } from "./sync-plan";

export class SyncPlanner {
  constructor(private readonly config: PlanningConfig) {}

  plan(existing: readonly ExistingLabel[]): SyncPlan {
    const entries: PlanEntry[] = [];
    const labelsByName = new Map(existing.map((label) => [labelKey(label.name), label]));
    const desiredKeys = new Set(this.config.labels.map((label) => labelKey(label.name)));

    for (const desired of this.config.labels) {
      const desiredKey = labelKey(desired.name);
      const current = labelsByName.get(desiredKey);
      const aliasMatches = uniqueLabels(
        desired.aliases
          .map((name) => labelsByName.get(labelKey(name)))
          .filter((label): label is ExistingLabel => label !== undefined),
      );

      if (current !== undefined) {
        if (labelChanged(current, desired)) {
          entries.push({ action: "update", name: current.name, desired });
        } else {
          entries.push({ action: "unchanged", name: current.name });
        }
        labelsByName.delete(labelKey(current.name));
        labelsByName.set(desiredKey, desired);

        for (const legacy of aliasMatches) {
          entries.push({ action: "delete", name: legacy.name, reason: "legacy_alias" });
          labelsByName.delete(labelKey(legacy.name));
        }
        continue;
      }

      if (aliasMatches.length > 1) {
        throw new Error(
          `多个旧标签同时映射到 ${desired.name}：${aliasMatches.map(({ name }) => name).join(", ")}`,
        );
      }

      const old = aliasMatches[0];
      if (old !== undefined) {
        entries.push({ action: "rename", name: old.name, desired });
        labelsByName.delete(labelKey(old.name));
        labelsByName.set(desiredKey, desired);
        continue;
      }

      entries.push({ action: "create", name: desired.name, desired });
      labelsByName.set(desiredKey, desired);
    }

    const remaining = [...labelsByName.values()].filter(
      (label) => !desiredKeys.has(labelKey(label.name)),
    );
    const staleManaged = remaining
      .filter((label) => this.config.managed(label.name))
      .sort(compareLabels);
    const repositorySpecific = remaining
      .filter((label) => !this.config.managed(label.name))
      .sort(compareLabels);

    for (const label of staleManaged) {
      entries.push({ action: "delete", name: label.name, reason: "stale_managed" });
    }
    for (const label of repositorySpecific) {
      entries.push({ action: "preserve", name: label.name });
    }

    return new SyncPlan(entries);
  }
}

function labelChanged(current: ExistingLabel, desired: LabelDefinition): boolean {
  return (
    current.name !== desired.name ||
    String(current.color).toUpperCase() !== desired.color ||
    String(current.description ?? "") !== desired.description
  );
}

function uniqueLabels(labels: readonly ExistingLabel[]): ExistingLabel[] {
  return [...new Map(labels.map((label) => [labelKey(label.name), label])).values()];
}

function compareLabels(left: ExistingLabel, right: ExistingLabel): number {
  return Buffer.compare(
    Buffer.from(left.name.toLowerCase(), "utf8"),
    Buffer.from(right.name.toLowerCase(), "utf8"),
  );
}
